compiler/rustc_parse/src/parser/mod.rs

```diff
@@ -1446,6 +1447,89 @@ impl<'a> Parser<'a> {
         Ok(())
     }
 
+    /// Parses `kw(in path)` plus shortcuts `kw(crate)` for `kw(in crate)`, `kw(self)` for
+    /// `kw(in self)` and `kw(super)` for `kw(in super)`.
+    fn parse_restriction(
+        &mut self,
+        kw: Symbol,
+        feature_gate: Option<Symbol>,
+        action: &str,
+        fbt: FollowedByType,
+    ) -> PResult<'a, Restriction> {
+        if !self.eat_keyword(kw) {
+            // We need a span, but there's inherently no keyword to grab a span from for an implied
+            // restriction. An empty span at the beginning of the current token is a reasonable
+            // fallback.
+            return Ok(Restriction::implied().with_span(self.token.span.shrink_to_lo()));
+        }
+
+        let gate = |span| {
+            if let Some(feature_gate) = feature_gate {
+                self.sess.gated_spans.gate(feature_gate, span);
+            }
+            span
+        };
+
+        let lo = self.prev_token.span;
+
+        self.expect(&token::OpenDelim(Delimiter::Parenthesis))?;
+
+        if self.check_keyword(kw::In) {
```

Comment:
Could this logic be shared between restriction and visibility parsing?

Reply:
I would imagine it could! Right now I'm more focused on fixing the known bug, but I'll definitely refactor after that.

Reply:
After making restrictions generic, it turns out to be nearly trivial to parse the entirety of visibilities as a restriction. Once it's parsed, I perform a simple conversion into the existing `Visibility` AST item.

In theory, I could fully replace `Visibility` with `Restriction<Visibility>`. I'm not going to do so as part of this PR, but it would be interesting to see.